src/gt_frontend/gtscript_to_gtir.py

```diff
@@ -0,0 +1,506 @@
+# -*- coding: utf-8 -*-
+#
+# Eve Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2020, CSCS - Swiss National Supercomputing Center, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import copy
+from typing import List, Union, cast
+
+import eve
+import gtc.common as common
+import gtc.unstructured.gtir as gtir
+
+from .built_in_types import BuiltInTypeMeta
+from .gtscript import Field, Local, Location, Mesh, TemporaryField
+from .gtscript_ast import (
+    Argument,
+    Assign,
+    BinaryOp,
+    Call,
+    Computation,
+    Constant,
+    Generator,
+    Interval,
+    IterationOrder,
+    LocationComprehension,
+    LocationSpecification,
+    Pass,
+    Stencil,
+    SubscriptMultiple,
+    SubscriptSingle,
+    Symbol,
+)
+
+
+_reduction_mapping = {
+    "sum": gtir.ReduceOperator.ADD,
+    "product": gtir.ReduceOperator.MUL,
+    "min": gtir.ReduceOperator.MIN,
+    "max": gtir.ReduceOperator.MAX,
+}
+
+# Types
+#  - BuiltInType
+#  - LocationType
+#  - DataType
+
+
+class SymbolTable:
+    def __init__(self, types, constants):
+        self.types = types
+        self.constants = constants
+
+    def __contains__(self, key):
+        return key in self.types
+
+    def __getitem__(self, item):
+        return self.types[item]
+
+    def __setitem__(self, item, val):
+        if item in self.types:
+            if (
+                self.types[item] == val
+            ):  # todo: just a workaround. remove when symbol table has proper scope!
+                return self.types[item]
+            raise ValueError("Symbol `{}` already in symbol table.".format(item))
+
+        self.types[item] = val
+        return self.types[item]
+
+    def materialize_constant(self, name, expected_type=None):
```

Comment:
A bit weird name for just a getter method with validation. I would use something simpler:
```suggestion
    def get_constant(self, name, expected_type=None):
```
Also the docstrings seems to be out of date

Reply:
The naming was intentional here to remind myself that this would currently be the place were constant folding would take place (for which I don't know yet what the right place is). It's just that there is nothing to "fold" since the values are externally supplied. I extended the docstring to emphasize this and added a todo.